doc/sphinx/proof-engine/tactics.rst

```diff
@@ -102,35 +101,54 @@ specified, the default selector is used.
 Bindings
 ~~~~~~~~
 
-Tactics that take a term as an argument may also accept :token:`bindings`
-to instantiate some parameters of the term by name or position.
-The general form of a term with :token:`bindings` is
-:n:`@term__tac with @bindings` where :token:`bindings` can take two different forms:
+Tactics that take a term as an argument may also accept :token:`bindings` to
+specify the values to assign unbound variables in a term.
+Bindings can be given by position or name.  Generally these appear in the form
+:n:`@one_term_with_bindings` or :n:`with @bindings`, depending on the tactic.
 
   .. insertprodn bindings bindings
 
   .. prodn::
-     bindings ::= {+ ( {| @ident | @natural } := @term ) }
-     | {+ @one_term }
+     bindings ::= {+ @one_term }
+     | {+ ( {| @ident | @natural } := @term ) }
 
-+ In the first form, if an :token:`ident` is specified, it must be bound in the
-  type of :n:`@term` and provides the tactic with an instance for the
-  parameter of this name. If a :token:`natural` is specified, it refers to
-  the ``n``-th non-dependent premise of :n:`@term__tac`.
+The first form binds free variables in the left-to-right order of their first
+appearance in the relevant term.  For some tactics, bindings for all free variables
+must be provided, such as for :tacn:`induction`, :tacn:`destruct`, :tacn:`elim`
+and :tacn:`case`.  Other tactics automatically generate some or all
+of the bindings from the conclusion or a hypothesis, such as :tacn:`apply` and
+:tacn:`constructor` and its variants.  In this case, only instances
+for the dependent products that are not bound in the conclusion of the relevant term
+are required (and permitted).
 
-  .. exn:: No such binder.
-     :undocumented:
+The second form binds variables by name (if :n:`@ident` is given), or
+by unifying with the ``n``-th :term:`premise` of the relevant term
+(if :n:`@natural` is given).
 
-+ In the second form, the interpretation of the :token:`one_term`\s depend on which
-  tactic they appear in.  For :tacn:`induction`, :tacn:`destruct`, :tacn:`elim`
-  and :tacn:`case`, the :token:`one_term`\s
-  provide instances for all the dependent products in the type of :n:`@term__tac` while in
-  the case of :tacn:`apply`, or of :tacn:`constructor` and its variants, only instances
-  for the dependent products that are not bound in the conclusion of :n:`@term__tac`
-  are required.
+   .. insertprodn one_term_with_bindings one_term_with_bindings
 
-  .. exn:: Not the right number of missing arguments.
-     :undocumented:
+   .. prodn::
+      one_term_with_bindings ::= @one_term {? with @bindings }
+
+In this form, bindings for variables in :n:`@one_term` are typically determined
+by unifying :n:`@one_term` with a tactic-dependent part of the context, with any
+remaining unbound variables provided by the :n:`bindings`.
```

Comment:
This explanation seems to contradict half of the "first form" paragraph, and indeed does not match what is described as the behavior for destruct. My suggestion would be to remove the paragraph entirely (it does not seem necessary to me) and to just insert `one_term_with_bindings` above `bindings` in the previous `prodn` (with no added explanation).

Reply:
"In this form" on line 133 refers to one_term_with_bindings.  It made perfect sense to me :-).

I think you'll like what I did.